hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantCreateRequest.java

```diff
@@ -0,0 +1,300 @@
+      OzoneManagerDoubleBufferHelper ozoneManagerDoubleBufferHelper) {
+
+    OMClientResponse omClientResponse = null;
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+    OmVolumeArgs omVolumeArgs;
+    boolean acquiredVolumeLock = false, acquiredUserLock = false;
+    Tenant tenant = null;
+    final String owner = getOmRequest().getUserInfo().getUserName();
+    LOG.info("owner: {}", owner);  // TODO: owner should be access_key_id
+    Map<String, String> auditMap = new HashMap<>();
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    CreateTenantRequest request = getOmRequest().getCreateTenantRequest();
+    final String tenantName = request.getTenantName();
+    final VolumeInfo volumeInfo =
+        getOmRequest().getCreateVolumeRequest().getVolumeInfo();
+    final String volumeName = volumeInfo.getVolume();
+    final String dbVolumeKey = omMetadataManager.getVolumeKey(volumeName);
+    IOException exception = null;
+    try {
+      // Check ACL: requires volume create permission. TODO: tenant create perm?
+      if (ozoneManager.getAclsEnabled()) {
+        checkAcls(ozoneManager, OzoneObj.ResourceType.VOLUME,
+            OzoneObj.StoreType.OZONE, IAccessAuthorizer.ACLType.CREATE,
+            tenantName, null, null);
+      }
+
+      acquiredVolumeLock = omMetadataManager.getLock().acquireWriteLock(
+          VOLUME_LOCK, volumeName);
+      // Check volume existence
+      if (omMetadataManager.getVolumeTable().isExist(volumeName)) {
+        LOG.debug("volume: {} already exists", volumeName);
+        throw new OMException("Volume already exists",
+            OMException.ResultCodes.VOLUME_ALREADY_EXISTS);
+      }
+      // Check tenant existence in tenantStateTable
+      if (omMetadataManager.getTenantStateTable().isExist(tenantName)) {
+        LOG.debug("tenant: {} already exists", tenantName);
+        throw new OMException("Tenant already exists",
+            OMException.ResultCodes.TENANT_ALREADY_EXISTS);
+      }
+
+      // Add to tenantStateTable. Redundant assignment for clarity
+      final String bucketNamespaceName = tenantName;
+      final String accountNamespaceName = tenantName;
+      final String userPolicyGroupName =
+          tenantName + OzoneConsts.DEFAULT_TENANT_USER_POLICY_SUFFIX;
+      final String bucketPolicyGroupName =
+          tenantName + OzoneConsts.DEFAULT_TENANT_BUCKET_POLICY_SUFFIX;
+      final OmDBTenantInfo omDBTenantInfo = new OmDBTenantInfo(
+          tenantName, bucketNamespaceName, accountNamespaceName,
+          userPolicyGroupName, bucketPolicyGroupName);
+      omMetadataManager.getTenantStateTable().addCacheEntry(
+          new CacheKey<>(tenantName),
+          new CacheValue<>(Optional.of(omDBTenantInfo), transactionLogIndex));
+
+      // Call OMMultiTenantManager
+//      tenant = ozoneManager.getMultiTenantManager().createTenant(tenantName);
```

Comment:
This would be a no go in actual (non prototype) code. We are making a Ranger call with volume lock, and from inside Ratis applyTxn.

Reply:
We can have a policy sync queue (and persistence) that pushes the policies to the external plugin in an async manner.

Reply:
@prashantpogde

Reply:
After a discussion, we think it should be better to move this chunk to `preExecute` in order to avoid writing to Ranger while holding a lock. But the (small) challenge then becomes how to properly clean up Ranger policy when the request fails at some point. CMIIW @avijayanhwx @prashantpogde

Reply:
Thanks for summarizing the discussion @smengcl.